### PR TITLE
Automation

### PR DESCRIPTION
**Pull request by MilosKozak**

### CODE REVIEW

**Comment by Heiner1** [comment]
Is there a chance to avouid submitting every profile change, when they are calculated but staying at the same level then before? It causes unreadable reports and may be data garbage.

File: app/src/main/java/info/nightscout/androidaps/services/LocationService.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+    @Override
+    public void onDestroy() {
+        if (L.isEnabled(L.LOCATION))
+            log.debug("onDestroy");
+        super.onDestroy();
+        MainApp.bus().unregister(this);
+        if (mLocationManager != null) {
+            try {
+                if (ActivityCompat.checkSelfPermission(this, android.Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, android.Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
```
Comment (adrian-hidothealth) [inline app/src/main/java/info/nightscout/androidaps/services/LocationService.java]: Shouldn't this rather be asked when we register for location updates and notify?
Or can this permission implicitly be revoked by toggling location on the phone?

Comment (AdrianLxM) [inline app/src/main/java/info/nightscout/androidaps/services/LocationService.java]: Shoudn't we check that when requesting location updates?
Or can they be implicitly revoked by toggling off location on the phone? If yes, shouldn't we still check earlier?

File: app/src/main/java/info/nightscout/androidaps/services/LocationService.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+                mLocationManager.requestLocationUpdates(
+                        LocationManager.NETWORK_PROVIDER,
+                        T.mins(5).msecs(),
+                        LOCATION_DISTANCE,
+                        mLocationListener = new LocationListener(LocationManager.NETWORK_PROVIDER)
+                );
+            if (SP.getString(R.string.key_location, "NONE").equals("GPS"))
+                mLocationManager.requestLocationUpdates(
+                        LocationManager.GPS_PROVIDER,
+                        T.mins(5).msecs(),
```
Comment (AdrianLxM) [inline app/src/main/java/info/nightscout/androidaps/services/LocationService.java]: I'd extract them as constants.

File: app/src/main/java/info/nightscout/androidaps/plugins/general/automation/AutomationPlugin.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+                .shortName(R.string.automation_short)
+                .preferencesId(R.xml.pref_automation)
+                .description(R.string.automation_description)
+        );
+    }
+
+    @Override
+    protected void onStart() {
+        Context context = MainApp.instance().getApplicationContext();
+        context.startService(new Intent(context, LocationService.class));
```
Comment (AdrianLxM) [inline app/src/main/java/info/nightscout/androidaps/plugins/general/automation/AutomationPlugin.java]: I don't have the full overview over the automation system yet. But wouldn't it be good to only start it if some component needs those updates?

File: app/src/main/java/info/nightscout/androidaps/services/LocationService.java (hunk 1 above)
Comment (MilosKozak) [reply app/src/main/java/info/nightscout/androidaps/services/LocationService.java]: I think ask when you enable Automation is ok. At this time location service is started

File: app/src/main/java/info/nightscout/androidaps/plugins/general/automation/AutomationPlugin.java (hunk 3 above)
Comment (MilosKozak) [reply app/src/main/java/info/nightscout/androidaps/plugins/general/automation/AutomationPlugin.java]: done

File: app/src/main/java/info/nightscout/androidaps/services/LocationService.java (hunk 2 above)
Comment (MilosKozak) [reply app/src/main/java/info/nightscout/androidaps/services/LocationService.java]: done

File: app/src/main/java/info/nightscout/androidaps/plugins/general/automation/AutomationPlugin.java (hunk 3 above)
Comment (MilosKozak) [reply app/src/main/java/info/nightscout/androidaps/plugins/general/automation/AutomationPlugin.java]: I'd not make it too complicated. In passive mode it doesn't cost more energy

**Pull request merged**